# nf-core/demultiplex: Output

## :warning: Please read this documentation on the nf-core website: [https://nf-co.re/demultiplex/output](https://nf-co.re/demultiplex/output)

> _Documentation of pipeline parameters is generated automatically from the pipeline schema and can no longer be found in markdown files._

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.


## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/)
and processes data using the following steps:

* Reformatting the input sample sheet to collapse [iCLIP](https://emea.illumina.com/science/sequencing-method-explorer/kits-and-arrays/iclip.html) samples into one lane and separate single cell 10X samples into separate sample sheets
* Checking the sample sheet for downstream error causing samples such as:
  * a mix of short and long indexes on the same lane
  * a mix of single and dual indexes on the same lane
* Processes that only run if there are issues within the sample sheet found by the sample sheet check process (CONDITIONAL):
  * Creates a new sample sheet with any samples that would cause an error removed and create a a txt file of a list of the removed problem samples
  * Run [`bcl2fastq`](http://emea.support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html) on the newly created sample sheet and output the Stats.json file
  * Parsing the Stats.json file for the indexes that were in the problem samples list.
  * Recheck newly made sample sheet for any errors or problem samples that did not match any indexes in the Stats.json file. If there is still an issue the pipeline will exit at this stage.
* [bcl2fastq](#bcl2fastq) - converting bcl files to fastq, and demultiplexing (CONDITIONAL)
* Processes that only run if there are 10X samples on the sample sheet input (CONDITIONAL):
  * [CellRanger](#cellranger) - demultiplexes raw base call (BCL) files generated by Illumina sequencers into FASTQ files and is a wrapper around Illumina's bcl2fastq
  * [CellRangerCount](#cellrangercount) - performs alignment, filtering, barcode counting, and UMI counting
* [FastQC](#fastqc) - read quality control
* [MultiQC](#multiqc) - aggregate report, describing results of the whole pipeline
* [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

## bcl2fastq

[bcl2fastq](http://emea.support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html) Conversion Software both demultiplexes data and converts BCL files generated by Illumina sequencing systems to standard FASTQ file formats for downstream analysis.

**Output directory: `params.outdir/projectName/FastQ`**

* `samplename/sample.fastq.gz`
  * Untrimmed raw fastq files

## CellRanger

[CellRanger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) is a set of analysis pipelines that process Chromium single-cell RNA-seq output to align reads, generate feature-barcode matrices and perform clustering and gene expression analysis. `cellranger mkfastq` demultiplexes raw base call (BCL) files generated by Illumina sequencers into FASTQ files and is a wrapper around Illumina's bcl2fastq. `cellranger count` takes FASTQ files from `cellranger mkfastq` and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis.

`cellranger mkfastq`

**Output directory: `params.outdir/mkfastq/outs/fastq_path/project_name/sample_name/`**

* `samplename/sample.fastq.gz`
  * Untrimmed raw fastq files

`cellranger count`

**Output directory: `params.outdir/count/projectName/sampleName/outs/`**

* `outs/analysis`
  * Secondary analysis data including dimensionality reduction, cell clustering, and differential expression
* `outs/web_summary.html`
  * Run summary metrics and charts in HTML format
* `outs/metrics_summary.csv`
  * Run summary metrics in CSV format

## FastQC

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences.

For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

**Output files:**

* `fastqc/`
  * `*_fastqc.html`: FastQC report containing quality metrics for your untrimmed raw fastq files.
* `fastqc/zips/`
  * `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.
* `sampleID`
  * Directory containing the FastQC report, tab-delimited data file and plot images

> **NB:** The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.

## MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarizing all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability.

For more information about how to use MultiQC reports, see [https://multiqc.info](https://multiqc.info).

**Output files:**

* `multiqc/`
  * `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  * `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  * `multiqc_plots/`: directory containing static images from the report in various formats.

## Pipeline information

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

**Output files:**

* `pipeline_info/`
  * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.csv`.
  * Documentation for interpretation of results in HTML format: `results_description.html`.
